### DISCUSSION TITLE
Error in post-command-hook (vertico--exhibit)

### DISCUSSION THREAD

**Discussion by Antigen-1** [minad/vertico#501] [Q&A] [answered]
```
$ emacs --version
GNU Emacs 29.
```

With a recent package update, vertico.el doesn't work now. Emacs says `Error in post-command-hook (vertico--exhibit): (void-function compat--completion-metadata-get)`.

**Comment by mukundzare** [comment]
I have this similar problem in Emacs 29:
`Error in pre-command-hook (vertico--prepare): (invalid-function static-if)`

My analysis: The macro `static-if` is defined in **compat-30.el** but can be found via `M-x describe-function` and it correctly points to the macro being defined in **compat-30.el**. Thinking that there could be a problem with the bytecode, I cleared all occurrences of **compat-30.el** as well as **compat-30.eln** and all the cached data. Reinstalled compat.el but the error won't resolve.

**Reply by minad** [reply]
This happens if Vertico was compiled against an old Compat version without the static-if macro. You have to recompile Vertico to fix this problem.

**Reply by mukundzare** [reply]
Okay, basic question: What do you mean by recompile?

**Reply by mukundzare** [reply, marked as answer]
Anyway, now I tried `package-delete vertico` and just reinstalled it from `list-packages` and it works. From my side this problem seems fixed.

**Reply by bgro** [reply]
For me, uninstalling and re-installing did not work to remove this problem -- I even reinstalled everything from scratch by removing `.emacs.d/elpa`. Currently I am working with an older version of packages (copying over the `elpa` folder from a machine  installed a few months ago, in which `vertico` still is dependent on `compat-29`.

I forgot to mention: this problem occurs for me both for Emacs 29.1 and Emacs 29.4.

**Reply by EXT-OWL** [reply]
> Anyway, now I tried `package-delete vertico` and just reinstalled it from `list-packages` and it works. From my side this problem seems fixed.

I can confirm this works, Using Emacs 30 on Windows 11 and this issue was driving me crazy.

**Reply by mentalisttraceur** [reply]
By the way, you can use `M-x package-recompile` from within Emacs (you can ignore the Vertico error in the prompts - just fully type `package-recompile` and `vertico`.) Or just directly eval `(package-recompile 'vertico)`.

That was enough to get it working for me, no need to fiddle with reinstalling.

**Reply by ruthra-kumar** [reply]
> By the way, you can use `M-x package-recompile` from within Emacs (you can ignore the Vertico error in the prompts - just fully type `package-recompile` and `vertico`.) Or just directly eval `(package-recompile 'vertico)`.
> 
> That was enough to get it working for me, no need to fiddle with reinstalling.

Recompiling worked for me.

emacs 30.1 on macos sequoia 15.3.1

**Reply by aeslaughter** [reply]
`M-x package-recompile RET vertico` also worked for me on 30.1

**Comment by kdmsnr** [comment]
Same error occurs with Emacs 30.0.50 and Vertico 1.9.

**Reply by minad** [reply]
Did you try to reinstall Compat/Vertico/Marginalia etc, see https://github.com/minad/vertico/discussions/501#discussioncomment-10058924?

**Comment by draxil** [comment]
Straight users:

I fixed this by going to ~/.emacs.d/straight/build and doing:

`rm -rf vertico* compat/ orderless/ consult* embark* marginalia/`

and re-start emacs and these things get force-rebuilt. May be a better way, but this worked for me.

**Reply by torfjelde** [reply]
Thanks; this worked:)

**Reply by hangyan** [reply]
trying emacs 30 pretest version and faced this. Thanks!

**Reply by titaniumbones** [reply]
For Elpaca:

> Straight users:
> 
> I fixed this by going to ~/.emacs.d/straight/build and doing:
> 
> `rm -rf vertico* compat/ orderless/ consult* embark* marginalia/`
> 
> and re-start emacs and these things get force-rebuilt. May be a better way, but this worked for me.

A very similar process fixed things for me in elpaca when upgrading to emacs-plus@30 using brew on macos:

```
cd ~/.config/emacs/elpaca
rm -r vertico embark embark-consult compat orderless citar-embark marginalia
```

thanks for hte help, this was kind of drivng me crazy.

**Reply by ljrtuit** [reply]
None of the previous suggestions in this post resolved my issue.
I had to do the following:

```
cd ~/.emacs.d/straight/repos
rm -r vertico embark embark-consult compat orderless citar-embark marginalia
```

**Comment by simendsjo** [comment]
Got this error when using Guix Home. Deleting the eln-cache resolved the issue.

**Reply by Basrandir** [reply]
This works. Unfortunately every time I use vertico it gets recreated and I need to delete it before starting a new Emacs instance.

**Reply by antler5** [reply]
Got this error on Guix when switching to emacs-next (30.0.91), only occurs on the second run when loading compiled elisp.

~~I was going to say I've downgraded to https://github.com/minad/vertico/commit/4498a2589c60cc788d6cb3909964c12c8fbad79ahttps://github.com/minad/vertico/commit/267d4c961592e8bd26e87938d0d65f072f5c5168https://github.com/minad/vertico/commit/4498a2589c60cc788d6cb3909964c12c8fbad79a but (after waiting for compilation to finish) that didn't help. I'm now figuring that the issue is triggered by the version-based conditional definitions in compat.el...~~
I don't think I downgraded successfully, just a sec~

Edit: Yep, the lastest commit was still getting pulled in as a dep -- fixed my package graph and am now holding over on https://github.com/minad/vertico/commit/4498a2589c60cc788d6cb3909964c12c8fbad79a.

**Reply by antler5** [reply]
> A minimal version of Compat will be present in Emacs version 30 and newer. Packages which are part of Emacs itself and want to take advantage of Compat, can also use (require 'compat). The advantage of the inclusion of a minimal Compat in Emacs is that Compat will not be installed if you require a version newer or equal than the current Emacs version. For example, if a package depending on Emacs 25.1 and Compat 29.1 is installed on Emacs 30.1, Compat will not be pulled in as dependency, since Emacs 30.1 already provides the required functionality.
> 
> Compat provides replacement functions with extended functionality for functions that are already defined, e.g., sort or assoc. These functions may have changed their calling convention (additional optional arguments) or may have changed their behavior. These functions must be looked up explicitly with compat-function or called explicitly with compat-call. We call them Extended Definitions. In contrast, newly Added Definitions can be called as usual. The Compat manual explicitly documents the calling convention of each compatibility function.
> -- https://elpa.gnu.org/packages/doc/compat.html#Usage-1

```elisp
;;; *scratch* -*- mode: emacs-lisp -*-
;; Error in pre-command-hook (vertico--prepare): (void-function compat--completion-metadata-get)
(let ((regex "compat--completion-metadata-get"))
  (->> (elisp-load-path-roots)
    (-filter #'file-exists-p)
    (mapcar #'shell-quote-argument)
    (append `("grep" "-R" ,(shell-quote-argument regex)))
    (funcall (-flip #'string-join) " ")
    (grep-find)))
```

<details>
<summary>Output</summary>
<pre>
grep: /gnu/store/152r147v6fz04zmzgaw7mbbhwbgrdjc3-profile/share/emacs/site-lisp/vertico-1.9/vertico-multiform.elc: binary file matches
grep: /gnu/store/152r147v6fz04zmzgaw7mbbhwbgrdjc3-profile/share/emacs/site-lisp/vertico-1.9/vertico.elc: binary file matches
grep: /gnu/store/152r147v6fz04zmzgaw7mbbhwbgrdjc3-profile/share/emacs/site-lisp/corfu-1.5/corfu.elc: binary file matches
grep: /gnu/store/152r147v6fz04zmzgaw7mbbhwbgrdjc3-profile/share/emacs/site-lisp/compat-30.0.0.0/compat-30.elc: binary file matches
grep: /gnu/store/152r147v6fz04zmzgaw7mbbhwbgrdjc3-profile/share/emacs/site-lisp/orderless-1.2/orderless-kwd.elc: binary file matches
grep: /gnu/store/152r147v6fz04zmzgaw7mbbhwbgrdjc3-profile/share/emacs/site-lisp/orderless-1.2/orderless.elc: binary file matches
grep: /gnu/store/152r147v6fz04zmzgaw7mbbhwbgrdjc3-profile/share/emacs/site-lisp/marginalia-1.7/marginalia.elc: binary file matches
grep: /gnu/store/5fz4hfsqgh2rgdz2gxpxsnf6ak2nbsnx-emacs-compat-30.0.0.0/share/emacs/site-lisp/compat-30.0.0.0/compat-30.elc: binary file matches
grep: /gnu/store/5m27zc0m6kjqkps997pypbqyzwga38jc-emacs-corfu-1.5/share/emacs/site-lisp/corfu-1.5/corfu.elc: binary file matches
grep: /gnu/store/220v7pr4m71bah8di483fx4qald03lab-emacs-marginalia-1.7/share/emacs/site-lisp/marginalia-1.7/marginalia.elc: binary file matches
grep: /gnu/store/5bsrm6nnpy74cgdca73z28kbyzbfjc82-emacs-vertico-1.9/share/emacs/site-lisp/vertico-1.9/vertico-multiform.elc: binary file matches
grep: /gnu/store/5bsrm6nnpy74cgdca73z28kbyzbfjc82-emacs-vertico-1.9/share/emacs/site-lisp/vertico-1.9/vertico.elc: binary file matches
grep: /gnu/store/x2s054z0hlygh4prvkvswvgpyglp9zkj-emacs-orderless-1.2/share/emacs/site-lisp/orderless-1.2/orderless-kwd.elc: binary file matches
grep: /gnu/store/x2s054z0hlygh4prvkvswvgpyglp9zkj-emacs-orderless-1.2/share/emacs/site-lisp/orderless-1.2/orderless.elc: binary file matches
</pre>
</details>

Hmm... I double checked that `emacs-next` has been correctly swapped into the dependency tree, but my hunch right now is that it's still being used in the build-system for pre-compilation of native byte-code.

**Reply by antler5** [reply]
```scheme
;;; scratch.scm -*- mode: scheme -*-
(module-set! (resolve-module '(gnu packages emacs))
  'emacs-minimal emacs-next-minimal)
```

```
Grep finished with no matches found at Sat Oct 19 10:13:21, duration 0.26 s
```

Yep, that's it (for me). Why does the error only crop up after first-load compilation if the issue is with package pre-compilation? No idea. I'll try updating the package-graph to use emacs-next-minimal and see if that reaches the build-system.

**Reply by antler5** [reply]
Last update: Any Guix user having this issue with `--with-input=emacs=emacs-next` (or a declarative equivalent) needs to pair that with`--with-input=emacs-minimal=emacs-next-minimal`, which may cause test failures for some packages (`emacs-lispy` IME) but will do for me.

**Comment by simendsjo** [comment]
If you are able to reproduce it, try reporting it to the Guix mailinglist worth the instructions.

**Comment by al3xandru** [comment]
I am seeing a similar, but not exactly the same error: `Error in post-command-hook (vertico--exhibit): (cl-assertion-failed lexical-binding)`. This happens with a nightly Emacs build: GNU Emacs 31.0.50 (build 1, aarch64-apple-darwin21.6.0, NS appkit-2113.65 Version 12.7.6 (Build 21H1320)) of 2025-02-02.  The version of vertico installed is 20250128.1729

I don't know how to investigate this further. If you have any hints I'd be happy to try to gather more details.


(Update: I forgot to mention that I did remove and reinstalled vertico. I also removed an uninstalled org-ql whose `org-ql-find` funciton is invoked when getting the error.  I also `package-recompile-all` just in case.)

**Comment by kalaleq** [comment]
Hi there, i've been trying to upgrade to emacs-30.0.x and today to emacs-30.1 with the release.

I'm getting a similar issue to the above but re-installing things doesn't seem to make any difference.

Mine is the original: `Error in post-command-hook (vertico--exhibit): (void-function compat--completion-metadata-get)`

Everything works fine on initial load, but once emacs finishes native byte-compiling things in the background, i get the above error.  Deleting the eln-cache directory starts the same process over again.

I saw that emacs-30 includes a 'stub' compat package.  I tried removing the separately installed compat package, same effect.  I tried removing the stub package and putting the "real" compat back in place, same effect.

I did notice that if, in emacs-30, i try `describe-function compat--completion-metadata-get`, i get the following:

`compat--completion-metadata-get is a null

[Missing arglist.]

Not documented.`

This seems suspicious.  When i try that in emacs-29 i get a standard function description referencing compat-30.el.

I did narrow things down to the compat-30.el library getting native byte-compiled.  If i remove just compat-30-....eln from the eln-cache directory, and run emacs, vertico works.  As soon as it gets compiled again, vertico doesn't work.

My understanding of debugging this issue has hit its limit, and i have to revert to emacs-29 for now.  But i'm hoping someone has a clue as to where i can look further...

**Reply by minad** [reply]
Remove Compat and reinstall Vertico, Marginalia, etc. I recommend to reinstall/recompile all packages when moving to a new Emacs version, since this way you end up with newly compiled byte code (possibly more efficient).

I had also mentioned here already that reinstalling the packages should fix any such issues: https://github.com/minad/vertico/discussions/501#discussioncomment-10262675

**Reply by kalaleq** [reply]
Ah!  Thank you so much for your kindness to randos.  I did actually try recompiling vertico itself at one point, but i'm no longer sure in combination with what other steps.

I've now simply re-compiled and re-installed all elisp packages as you suggested and everything is working smoothly.  I love all your emacs work, by the way.

**Comment by IceAsteroid** [comment]
Installed 30.1 from 29.x version. I had a same issue, deleting the `elpa-cache` and the `vertico` package in `elpa` directory didn't work ,but after deleting all packages in `elpa` and reinstalling worked

**Comment by thefourtheye** [comment]
I also faced the same problem, with `GNU Emacs 30.1`. `rm -rf ~/.emacs.d/elpa/vertico-*` helped me resolve this. I had to reinstall `vertico` after deleting those directories.

I couldn't `package-delete vertico`, as the prompt said `[No match found]` or something similar, when I typed `vertico` and pressed Enter.

**Comment by jbmestelan** [comment]
None of these worked for me (but thanks for the leads). What eventually worked was removing the byte-compiled files from `~/.cache/emacs` ; after noticing I have `(setq user-emacs-directory "~/.cache/emacs")`

```
cd ~/.cache/emacs/eln-cache/30.1-c7a97098
rm -fr consult-* vertico-* embar* embark-consult compat-* marginalia-*
```
